### PR TITLE
Bitpacking

### PR DESCRIPTION
**Pull request by vayuda**
Based on this issue:  #284 

Adding this first iteration of packing/unpacking algorithms to support lower bit dtypes into protoype/

### CODE REVIEW

File: torchao/prototype/common/bitpacking.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        padding = torch.zeros((*data.shape[:-1], scale - data.shape[-1] % scale), dtype=data.dtype).cuda()
+        data = torch.cat([data, padding], dim=-1).cuda()
+    
+    shape = data.shape
+    data = data.contiguous().view(-1)
+    #shift the data to the different indexes within the larger dtype and then union them together
+    ret = reduce(lambda x,y: x|y,[data[i::scale] << container_size-data_size*(i+1) for i in range(scale)]).cuda()
+    return ret.view(*shape[:-1],shape[-1] // scale)
+
+if __name__ == '__main__':
```
Comment (msaroufim) [inline torchao/prototype/common/bitpacking.py:103]: Can you move all of the below to the test folder https://github.com/pytorch/ao/tree/main/test/quantization

Call it something like `test_bitpack.py`

File: test/prototype/test_bitpacking.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    unpacked = unpack(packed, 4)
+    unpadded = unpacked[:test_tensor.shape[0], ...]
+    assert(unpadded.allclose(test_tensor))
+    
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(not has_triton(), reason="unsupported without triton")
+def test_uint4_to_uint8_compile():
+    torch._dynamo.config.specialize_int = True
+    pack_compiled = torch.compile(pack)
+    unpack_compiled = torch.compile(unpack)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:21]: can you test `torch.compile(unpack, fullgraph=True)` that'll ensure only 1 kernel is ever generated

File: test/prototype/test_bitpacking.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    unpacked = unpack(packed, 4)
+    unpadded = unpacked[:test_tensor.shape[0], ...]
+    assert(unpadded.allclose(test_tensor))
+    
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(not has_triton(), reason="unsupported without triton")
+def test_uint4_to_uint8_compile():
+    torch._dynamo.config.specialize_int = True
+    pack_compiled = torch.compile(pack)
+    unpack_compiled = torch.compile(unpack, full_graph=True)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:21]: it's fullgraph=True

Comment (msaroufim) [inline test/prototype/test_bitpacking.py:21]:
```suggestion
    unpack_compiled = torch.compile(unpack, fullgraph=True)
```

File: test/prototype/test_bitpacking.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    unpacked = unpack(packed, 3)
+    unpadded = unpacked[:test_tensor.shape[0], ...]
+    assert(unpadded.allclose(test_tensor))
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(not has_triton(), reason="unsupported without triton")
+def test_uint2_to_uint8_col_wise_compile():
+    torch._dynamo.config.specialize_int = True
+    pack_compiled = torch.compile(pack)
+    unpack_compiled = torch.compile(unpack, full_graph=True)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:40]:
```suggestion
    unpack_compiled = torch.compile(unpack, fullgraph=True)
```

File: test/prototype/test_bitpacking.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    packed = pack(test_tensor, 8, 4)
+    unpacked = unpack(packed, 4)
+    unpadded = unpacked[:test_tensor.shape[0], ...]
+    assert(unpadded.allclose(test_tensor))
+    
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(not has_triton(), reason="unsupported without triton")
+def test_uint4_to_uint8_compile():
+    torch._dynamo.config.specialize_int = True
+    pack_compiled = torch.compile(pack)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:20]:
```suggestion
    pack_compiled = torch.compile(pack, fullgraph=True)
```

File: test/prototype/test_bitpacking.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    packed = pack(test_tensor,16, 3)
+    unpacked = unpack(packed, 3)
+    unpadded = unpacked[:test_tensor.shape[0], ...]
+    assert(unpadded.allclose(test_tensor))
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.skipif(not has_triton(), reason="unsupported without triton")
+def test_uint2_to_uint8_col_wise_compile():
+    torch._dynamo.config.specialize_int = True
+    pack_compiled = torch.compile(pack)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:39]:
```suggestion
    pack_compiled = torch.compile(pack, fullgraph=True)
```

File: torchao/prototype/common/bitpacking.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        return _unpack_by_cols(data, data_size)
+    
+def pack(data, container_size, data_size, by_rows = True, device="cuda"):
+    """
+    Packs small dtype elements into a larger dtype.
+    Pads rows to be divisible by the scale.
+    
+    Inputs:
+    data: torch.Tensor - a tensor of unpacked elements of a small dtype.
+    container_size: int - the size of the large dtype in bits.
```
Comment (gau-nernst) [inline torchao/prototype/common/bitpacking.py:36]: Just curious. `container_size` can be determined from `data.dtype` right? e.g. uint8 -> 8, uint16 -> 16. (there is also this - https://pytorch.org/docs/stable/type_info.html#torch.torch.iinfo).
Also, is it assumed that `data.dtype` has `container_size` number of bits? What if `data` use larger or smaller bit-width than `container_size`? e.g. store int4 in int32, then request to pack to int8. Depending on what are your assumptions to the inputs, perhaps some kind of type checking and/or type casting is good.

**Review by msaroufim — APPROVED** [review]
Nice first step let's keep iterating on cuda mode to figure out how to promote this to a stable feature

**Pull request merged**